packages/auth/src/grant/routes.ts

```diff
@@ -84,3 +94,82 @@ async function createGrantInitiation(
   ctx.status = 201
   ctx.body = res
 }
+
+async function startInteraction(
+  deps: ServiceDependencies,
+  ctx: AppContext
+): Promise<void> {
+  const { interactId } = ctx.params
+  const { config, grantService, clientService } = deps
+  const grant = await grantService.getByInteraction(interactId)
+
+  if (!grant) {
+    ctx.status = 401
+    ctx.body = {
+      error: 'unknown_request'
+    }
+
+    return
+  }
+
+  ctx.session.interactId = grant.interactId
+
+  const registryData = await clientService.getRegistryDataByKid(
+    grant.clientKeyId
+  )
+
+  if (!registryData) {
+    ctx.status = 401
+    ctx.body = {
+      error: 'invalid_client'
+    }
+    return
+  }
+
+  const interactionUrl = new URL(config.identityServerDomain)
+  interactionUrl.searchParams.set('clientName', registryData.name)
+  interactionUrl.searchParams.set('clientUri', registryData.url)
+
+  ctx.redirect(interactionUrl.toString())
+}
+
+async function finishInteraction(
+  deps: ServiceDependencies,
+  ctx: AppContext
+): Promise<void> {
+  const { interactId } = ctx.params
+  const interactSession = ctx.session.interactId
+
+  if (!interactSession || !interactId || interactSession !== interactId) {
+    ctx.status = 401
+    ctx.body = {
+      error: 'invalid_request'
+    }
+    return
+  }
+
+  const { grantService, config } = deps
+  const grant = await grantService.getByInteraction(interactId)
+
+  if (!grant) {
+    ctx.status = 404
+    ctx.body = {
+      error: 'unknown_request'
+    }
+    return
+  }
+
+  await grantService.issueGrant(grant.id)
+
+  const clientRedirectUri = new URL(grant.finishUri)
+  const { clientNonce, interactNonce, interactRef } = grant
+  const interactUrl = config.identityServerDomain + `/interact/${interactId}`
+
+  // https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol#section-4.2.3
+  const data = `${clientNonce}\n${interactNonce}\n${interactRef}\n${interactUrl}`
+
+  const hash = crypto.createHash('sha3-512').update(data).digest('base64')
+  clientRedirectUri.searchParams.set('hash', hash)
+  clientRedirectUri.searchParams.set('interact_ref', interactRef)
+  ctx.redirect(clientRedirectUri.toString())
```

Comment:
I guess I'm a bit confused about the user redirect in general. 
`start` will redirect the user to the identity server. There, the user logs in and (for now) accepts the grant. The identity server does a POST to the `finish` endpoint to confirm that the user accepted the grant. Wouldn't the identity server have to redirect the user to the `clientRedirectURI`?

Reply:
Yes, it's the [RO's job to redirect the user](https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol#section-4.2.1). I think unless the RO made this POST request in a `<form>` element, it would end up being the one performing the redirect in the `Location` header of this response.

Reply:
Got it, thanks!